### PR TITLE
[PBM-1143] Add support for WebDAV backups

### PR DESCRIPTION
**Pull request by dalbani**
https://jira.percona.com/browse/PBM-1143

### CODE REVIEW

File: pbm/storage/webdav/webdav.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	"io"
+	"os"
+	"path"
+	"strings"
+)
+
+type Conf struct {
+	ServerURL   string      `bson:"serverUrl,omitempty" json:"serverUrl" yaml:"serverUrl,omitempty"`
+	Prefix      string      `bson:"prefix,omitempty" json:"prefix,omitempty" yaml:"prefix,omitempty"`
+	Credentials Credentials `bson:"credentials" json:"-" yaml:"credentials"`
```
Comment (dalbani) [inline pbm/storage/webdav/webdav.go:17]: I've copied this over from `s3.go` and `azure.go`.
But I'm wondering, is there a reason why it's set to `-` only for `json`?

Comment (defbin) [reply pbm/storage/webdav/webdav.go:17]: Not sure why. I guess it to not be shown on debug output. The `bson:` is required to be stored in MongoDB.

Comment (dalbani) [reply pbm/storage/webdav/webdav.go:17]: 👍

**Comment by dalbani** [comment]
What's missing to this PR is maybe an integration test? Maybe it's not obvious to me how to do that.

**Comment by defbin** [comment]
try to add it here https://github.com/percona/percona-backup-mongodb/blob/86d552880b047573b1be93de493186b266d60f75/e2e-tests/cmd/pbm-test/run.go#L28-L31

**Comment by dalbani** [comment]
OK, then I need to fire up a WebDAV server, as I can rely on a external service like S3 or Azure.

**Comment by defbin** [comment]
to run the test:
`env MONGODB_VERSION=6.0 TESTS_BCP_TYPE=logical e2e-tests/run-all`

on Apple Silicon you can try:
`env MONGODB_VERSION=6.0.6-arm64 TESTS_BCP_TYPE= logical e2e-tests/run-all`

**Comment by dalbani** [comment]
Although it's all green, I am not sure if what I pushed actually run the E2E test for WebDAV 🤨

Speaking of E2E setup, I'm wondering if this choice of `path-ignores` is correct: https://github.com/percona/percona-backup-mongodb/blob/4ed23327208d9de4f01384e25911ceb14bd03686/.github/workflows/ci.yml#L22

In the sense that:
> When all the path names match patterns in paths-ignore, the workflow will not run. If any path names do not match patterns in paths-ignore, even if some path names match the patterns, the workflow will run.

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-paths

For example, I only committed to `e2e-tests/pkg/tests/sharded/cluster.go` in https://github.com/percona/percona-backup-mongodb/pull/855/commits/96b095ea952616a3e6efd2f0ddb205b88e260b47, and, yet, CI ran.

**Comment by capri1989** [comment]
Hi @dalbani,

just for me to better understand the use-case of WebDAV in 2023. In which scenario and where do you actually need WebDAV? Can you elaborate a little more on your needs?

Thanks,
Kai

**Comment by dalbani** [comment]
Well, even though it's 2023, WebDAV is very much "alive".
I can name for example two storage providers which offer WebDAV-backed repositories — and no S3 compatible endpoint:
- [Hetzner Storage Box](https://docs.hetzner.com/robot/storage-box/)
- Microsoft OneDrive — yep, it's technically possible to connect via WebDAV, you can look it up 😃 

And there are other (open source) implementations of WebDAV servers available for self-hosted deployments.

**Comment by capri1989** [comment]
Hey,

we've investigated and surveyed our current customer base, and there isn't genuine interest in WebDAV. Given that we need to maintain the code for a very long time, given support and SLAs, we don't consider adding it to the project.

So I'll close this PR, so you don't wait any longer on a decision that might never happen. Nevertheless thanks a lot for the contribution.

**Comment by dalbani** [comment]
Thanks for your feedback.
In the end, I've come to the conclusion that *logical* backups are not suited to my use case.
Due to the size of my database, the generated backup files were like 200+ GB (!), even with the best compression.
So I need to have a look at physical backups, is my understanding.

**Pull request closed without merging**